# [한국철도공사] 광역+여객 일별_역별_시간대별 실적

## 1. Data set description
여객(train), 광역(subway) 두 개의 형태로 나누어져 있으며,   
여객은 KTX 고속열차, 새마을호, 무궁화호 등 간선여객열차의 데이터, 광역은 전철 데이터이다.  
각각 2019.01.01 ~ 2019.05.31 / 2020.01.01 ~ 2020.05.31와 같은 두 개의 기간별 데이터로 구성  (따라서 전체 데이터셋은 4개)
1. KR_TB_TR_STN_DAY_CON~ : train, 기차역 중심 데이터
2. KB_TB_SBWY_STN_HR_CON~ : subway, 전철 중심 데이터


### Pakages import

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
plt.style.use('seaborn')

In [3]:
pd.set_option('max_columns', 24,'max_rows', 20) 
# pd.set_option('display.max_rows', None)

plt.style.use('seaborn')
plt.rc('font', family='Malgun Gothic')
plt.rc('axes', unicode_minus=False)

In [4]:
train19 = pd.read_csv("여객/KR_TB_TR_STN_DAY_CON_20190101_20190531_UTF8.csv", encoding='utf-8')
train20 = pd.read_csv("여객/KR_TB_TR_STN_DAY_CON_20200101_20200531_UTF8.csv", encoding='utf-8')
sub19 = pd.read_csv("광역/KR_TB_SBWY_STN_HR_CON_20190101_20190531_UTF8.csv", encoding='utf-8')
sub20 = pd.read_csv("광역/KR_TB_SBWY_STN_HR_CON_20200101_20200531_UTF8.csv", encoding='utf-8')

## 2. Variable Description

#### KR_TB_TR_STN_DAY_CON :: TRAIN 데이터
1. RUN_DT : 운행일자 / YYYYMMDD / CHAR
2. STN_CD : 역코드 / CHAR
3. STN_CD_NM : 역코드명 / 역명 / VARCHAR2
4. ABRD_PRNB : 승차인원수 / 해당역에서 승차한 총 인원수 / NUMBER
5. GOFF_PRNB : 하차인원수 / 해당역에서 하차한 총 인원수 / NUMBER

2019년 36,718개   / 2020년 35,040개  

In [5]:
train19

RUN_DT   STN_CD STN_CD_NM  ABRD_PRNB   GOFF_PRNB 
0      20190101      924       송도교          1           - 
1      20190101  3900023        서울     35,964      53,079 
2      20190101  3900025        용산     19,207      26,238 
3      20190101  3900030       영등포     11,573      11,538 
4      20190101  3900039        안양        415         348 
...         ...      ...       ...         ...         ...
36713  20190531  3901063    인천공항T2        154         270 
36714  20190531  3901065        월포         14          44 
36715  20190531  3901066        장사         17          18 
36716  20190531  3901067        강구         19          29 
36717  20190531  3901068        영덕        144         171 

[36718 rows x 5 columns]

In [6]:
train20

RUN_DT   STN_CD STN_CD_NM  ABRD_PRNB   GOFF_PRNB 
0      20200101      924       송도교          2           - 
1      20200101  3900023        서울     38,306      50,440 
2      20200101  3900025        용산     19,702      23,967 
3      20200101  3900030       영등포     10,812      11,125 
4      20200101  3900039        안양        387         269 
...         ...      ...       ...         ...         ...
35035  20200531  3901063    인천공항T2          1           - 
35036  20200531  3901065        월포         31          14 
35037  20200531  3901066        장사          5           6 
35038  20200531  3901067        강구         33           9 
35039  20200531  3901068        영덕        108         119 

[35040 rows x 5 columns]

#### KR_TB_SBWY_STN_HR_CON :: 전철 데이터
1. RUN_DT : 운행일자 / YYYYMMDD / CHAR
2. SBWY_STN_PRPR_NO : 전쳘역고유번호 / 전쳘역 관리코드 / VARCHAR2
3. SBWY_STN_CD_NM : 전철역코드명 / 전철역명 / VARCHAR2
4. TMWD_DV : 시간대구분 / 00시~23시까지 시간대 표시 / CHAR
5. SBWY_ABRD_PRNB : 전철승차인원수 / 해당역에서 승차한 총 인원수 / NUMBER
6. SBWY_GOFF_PRNB : 전철하차인원수 / 해당역에서 하차한 총 인원수 / NUMBER

2019년 857,876개 / 2020년 860,702개  
로 여객 데이터 수보다 훨씬 많은 것을 볼 수 있음. (아무래도 시간대별로 수집했으므로)

In [7]:
sub19

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV SBWY_ABRD_PRNB  \
0       20190101              1001             서울     00시             7    
1       20190101              1001             서울     05시            10    
2       20190101              1001             서울     06시            35    
3       20190101              1001             서울     07시            31    
4       20190101              1001             서울     08시            43    
...          ...               ...            ...     ...            ...   
857871  20190531              8014             일광     19시            95    
857872  20190531              8014             일광     20시            45    
857873  20190531              8014             일광     21시            70    
857874  20190531              8014             일광     22시            54    
857875  20190531              8014             일광     23시            50    

       SBWY_GOFF_PRNB  
0                 22   
1                  1   
2                  4   
3                 39   
4                101   
...               ...  
857871           109   
857872            73   
857873            57   
857874            29   
857875             6   

[857876 rows x 6 columns]

In [8]:
sub20

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV SBWY_ABRD_PRNB  \
0       20200101              1001             서울     00시             4    
1       20200101              1001             서울     05시            13    
2       20200101              1001             서울     06시            59    
3       20200101              1001             서울     07시            32    
4       20200101              1001             서울     08시            61    
...          ...               ...            ...     ...            ...   
860707  20200531              8015           부산원동     19시            62    
860708  20200531              8015           부산원동     20시            41    
860709  20200531              8015           부산원동     21시            56    
860710  20200531              8015           부산원동     22시            13    
860711  20200531              8015           부산원동     23시            10    

       SBWY_GOFF_PRNB  
0                 13   
1                  1   
2                  5   
3                 70   
4                113   
...               ...  
860707            24   
860708            18   
860709            12   
860710             -   
860711             5   

[860712 rows x 6 columns]

## 3. KR_TB_TR_STN_DAY_CON :: TRAIN 데이터

### 3-1. 운행일자(RUN_DT)

In [9]:
run19 = train19['RUN_DT'].value_counts()
run20 = train20['RUN_DT'].value_counts()

In [11]:
run19
#run19/20은 이렇게 날짜별로 각 몇개의 데이터가 수집되어 있는가에 대한 정보이다.

20190220    253
20190126    252
20190221    252
20190102    252
20190131    252
           ... 
20190416    233
20190409    232
20190513    232
20190506    232
20190430    231
Name: RUN_DT, Length: 151, dtype: int64

In [12]:
run19.describe()

count    151.000000
mean     243.165563
std        5.439890
min      231.000000
25%      239.000000
50%      243.000000
75%      248.000000
max      253.000000
Name: RUN_DT, dtype: float64

In [13]:
run20.describe()

count    152.000000
mean     230.526316
std        3.269258
min      223.000000
25%      228.000000
50%      230.000000
75%      233.000000
max      239.000000
Name: RUN_DT, dtype: float64

① 2019년 데이터 ::  
2019.01.01 ~ 2019.05.31 사이 총 151개의 날짜  (매일운행)  
min, mean, max가 순서대로 231, 243, 253인 것으로 보아 날짜별로 데이터의 차이는 없음.  
각 날짜별로 평균 243개의 데이터가 기록되어 있다고 볼 수 있다.

② 2020년 데이터 ::  
2020.01.01 ~ 2020.05.31 사이 총 152개의 날짜  (매일운행)  
min, mean, max가 순서대로 223, 230, 239인 것으로 보아 날짜별로 데이터의 차이는 없음.  
각 날짜별로 평균 230개의 데이터가 기록되어 있다고 볼 수 있다.

### 3-2. 역 코드명 (STN_CD_NM) // 역 코드 (STN_CD)

In [16]:
stn_nm19

임실      151
대전      151
의성      151
횡성      151
판교      151
       ... 
별내        2
망우        2
의왕        1
자미원       1
망상해변      1
Name: STN_CD_NM, Length: 266, dtype: int64

In [15]:
stn_nm19 = train19['STN_CD_NM'].value_counts()
stn_nm20 = train20['STN_CD_NM'].value_counts()

In [12]:
print(stn_nm19.index.unique()) #2019년 데이터에서 역은 중복없이 266개
print(stn_nm20.index.unique()) #2020년 데이터에서 역은 중복없이 243개

Index(['김천구미', '평창', '진주', '구미', '철암', '서광주', '중리', '대천', '왕십리', '삽교',
       ...
       '운천', '서빙고', '백운', '광운대', '창동', '망우', '별내', '자미원', '의왕', '망상해변'],
      dtype='object', length=266)
Index(['김천구미', '진주', '서광주', '조성', '왕십리', '삽교', '남창', '안동', '덕하', '곡성',
       ...
       '백양리', '정선', '송도교', '아우라지', '나전', '별어곡', '선평', '비동', '광운대', '동두천'],
      dtype='object', length=243)


2019, 2020년도 각각 266개, 243개의 지하철역을 대상으로 수집된 것을 알 수 있다.

In [13]:
stn_nm19.sort_values() #빈도수로 오름차순 정렬

망상해변        1
의왕          1
자미원         1
별내          2
망우          2
         ... 
구포        151
불국사       151
연산        151
인천공항T1    151
강구        151
Name: STN_CD_NM, Length: 266, dtype: int64

In [14]:
stn_nm20.sort_values() #빈도수로 오름차순 정렬

동두천      1
광운대      1
비동       3
선평       8
별어곡     12
      ... 
구포     152
불국사    152
만종     152
진영     152
진상     152
Name: STN_CD_NM, Length: 243, dtype: int64

모든 역에서 매일매일 데이터가 수집되진 않은 것 같다. 

In [15]:
#예를 들어, 망우역의 경우 2019년에서 2개의 데이터가 있었으며 각각 2월 24일, 4월 20일이다.
train19[train19['STN_CD_NM'] == "망우"]

RUN_DT   STN_CD STN_CD_NM  ABRD_PRNB   GOFF_PRNB 
13373  20190224  3900116        망우        103         103 
26764  20190420  3900116        망우        224         224

### 3-3. 승/하차 인원수 (ABRD_PRNB/GOFF_PRNB)

In [16]:
train19.columns #컬럼 확인, 스페이스는 왜 들어간거?

Index(['RUN_DT', 'STN_CD', 'STN_CD_NM', ' ABRD_PRNB ', ' GOFF_PRNB '], dtype='object')

In [17]:
train19[' ABRD_PRNB '] 

0              1 
1         35,964 
2         19,207 
3         11,573 
4            415 
           ...   
36713        154 
36714         14 
36715         17 
36716         19 
36717        144 
Name:  ABRD_PRNB , Length: 36718, dtype: object

In [18]:
train19[' GOFF_PRNB ']

0              - 
1         53,079 
2         26,238 
3         11,538 
4            348 
           ...   
36713        270 
36714         44 
36715         18 
36716         29 
36717        171 
Name:  GOFF_PRNB , Length: 36718, dtype: object

describe()를 찍어서 min, mean, max 값을 확인해보려고 했지만, "," 때문에 문자로 인식되어서 할 수 없다.   
따라서 아래의 전처리를 통해 숫자형으로 바꾸도록 한다.

In [19]:
def prnb_to_int(bf_series):
    series = bf_series.copy()
    
    #replace ',' -> ''
    for i in range(len(series)):
        series[i] = series[i].replace(',', '')
    
    #check missing values
    print("' - ' count : ", series[series == ' - '].count())
    print("'0' count : ", series[series == '0'].count())
    
    #replace ' - ' -> '0'
    for i in range(len(series)):
        series[i] = series[i].replace(' - ', '0')
    
    #re-check missing values
    print("after modification '0' count : ", series[series == '0'].count())

    #pirnt a describe result
    series = series.astype(str).astype(int).astype("Int64")
    
    for i in range(len(series)):
        if series[i] == 0:
            series[i] = np.nan
            
    print("after modification 0 count : ", series[series == 0].count())
    print("after modification 'nan' count : ", series.isnull().sum())
    print(series.describe())
    
    return series

In [20]:
print("\n**2019년도 여객 승차 데이터**")
train19[' ABRD_PRNB '] = prnb_to_int(train19[' ABRD_PRNB '])
print("\n**2019년도 여객 하차 데이터**")
train19[' GOFF_PRNB '] = prnb_to_int(train19[' GOFF_PRNB '])

print("")

print("\n**2020년도 여객 승차 데이터**")
train20[' ABRD_PRNB '] = prnb_to_int(train20[' ABRD_PRNB '])
print("\n**2020년도 여객 하차 데이터**")
train20[' GOFF_PRNB '] = prnb_to_int(train20[' GOFF_PRNB '])


**2019년도 여객 승차 데이터**
' - ' count :  360
'0' count :  0
after modification '0' count :  360
after modification 0 count :  0
after modification 'nan' count :  360
count    36358.000000
mean      1644.254222
std       4908.589764
min          1.000000
25%         31.000000
50%        162.000000
75%       1003.000000
max      77357.000000
Name:  ABRD_PRNB , dtype: float64

**2019년도 여객 하차 데이터**
' - ' count :  551
'0' count :  0
after modification '0' count :  551
after modification 0 count :  0
after modification 'nan' count :  551
count    36167.000000
mean      1652.937623
std       4966.229391
min          1.000000
25%         30.000000
50%        167.000000
75%       1003.000000
max      80464.000000
Name:  GOFF_PRNB , dtype: float64


**2020년도 여객 승차 데이터**
' - ' count :  288
'0' count :  0
after modification '0' count :  288
after modification 0 count :  0
after modification 'nan' count :  288
count    34752.000000
mean      1053.389762
std       3268.873079
min          1.000000
25%  

모두 비슷한 값을 가지고 있는 것을 볼 수 있다.   
다만 모든 수치에서 2020년도 자료가 더 적게 나왔다. (코로나의 영향?)

각 데이터에서 max 값을 가지고 있는 역은 무엇인지 찍어보았다.

In [21]:
max(train19[train19[' ABRD_PRNB '] != np.nan][' ABRD_PRNB '])

77357

In [22]:
# 2019년 철도 승차인원 MAX == "서울역"
train19[train19[' ABRD_PRNB '] == max(train19[train19[' ABRD_PRNB '] != np.nan][' ABRD_PRNB '])]

RUN_DT   STN_CD STN_CD_NM   ABRD_PRNB    GOFF_PRNB 
30053  20190504  3900023        서울        77357        63603

In [23]:
# 2019년 철도 하차인원 MAX == "서울역"
train19[train19[' GOFF_PRNB '] == max(train19[train19[' GOFF_PRNB '] != np.nan][' GOFF_PRNB '])]

RUN_DT   STN_CD STN_CD_NM   ABRD_PRNB    GOFF_PRNB 
8872  20190206  3900023        서울        35353        80464

In [24]:
# 2019년 철도 승차인원 MAX == "서울역"
train20[train20[' ABRD_PRNB '] == max(train20[train20[' ABRD_PRNB '] != np.nan][' ABRD_PRNB '])]

RUN_DT   STN_CD STN_CD_NM   ABRD_PRNB    GOFF_PRNB 
5168  20200123  3900023        서울        77307        41397

In [25]:
# 2019년 철도 승차인원 MAX == "서울역"
train20[train20[' GOFF_PRNB '] == max(train20[train20[' GOFF_PRNB '] != np.nan][' GOFF_PRNB '])]

RUN_DT   STN_CD STN_CD_NM   ABRD_PRNB    GOFF_PRNB 
5874  20200126  3900023        서울        36857        73680

모든 연도별/승하차 데이터에서 서울역이 1등으로 가장 많았다. 

-------------

## 4. KR_TB_TR_STN_DAY_CON :: SUBWAY 데이터

### 4-1. RUN_DT (운행일자)

In [26]:
run19 = sub19['RUN_DT'].value_counts()
run20 = sub20['RUN_DT'].value_counts()

In [27]:
run19.describe()

count     151.000000
mean     5681.298013
std        54.552946
min      5575.000000
25%      5617.000000
50%      5704.000000
75%      5721.000000
max      5759.000000
Name: RUN_DT, dtype: float64

In [28]:
run20.describe()

count     152.000000
mean     5662.578947
std        41.255708
min      5568.000000
25%      5633.250000
50%      5668.500000
75%      5696.250000
max      5761.000000
Name: RUN_DT, dtype: float64

지하철 데이터도 마찬가지로 2019년에서 151개, 2020년에서 152개가 관측된 것을 볼 수 있으며,  
데이터 수가 월등히 많았기 때문에 아까 기차 데이터에서 일자별로 평균 240개 정도가 있던 것과는 다르게 일별 5600개 정도의 데이터가 있는 것을 볼 수 있다.

### 4-2. 전철역고유번호(SBWY_STN_PRPR_NO) // 전쳘역코드명(SBWY_STN_CD_NM)

In [29]:
stn_nm19 = sub19['SBWY_STN_CD_NM'].value_counts()
stn_nm20 = sub20['SBWY_STN_CD_NM'].value_counts()

In [30]:
print(stn_nm19.index.unique())
print(stn_nm20.index.unique())

Index(['부평', '영등포', '안산', '수원', '부천', '안양', '산본', '금정', '왕십리', '회기',
       ...
       '신원', '광명', '대성리', '신촌', '서울(경의선)', '굴봉산', '백양리', '달월', '옥수', '지평'],
      dtype='object', length=274)
Index(['부평', '안산', '정자', '영등포', '부천', '서현', '주안', '금정', '왕십리', '삼송',
       ...
       '신원', '신촌', '대성리', '백양리', '굴봉산', '달월', '옥수', '지평', '부산원동', '임진강'],
      dtype='object', length=276)


2019년 데이터에는 총 274개의 전철역이,   
2020년 데이터에는 총 276개의 전철역이 있는 것을 알 수 있다.

In [31]:
print(len(sub19['TMWD_DV'].unique()))
sub19['TMWD_DV'].unique()

24


array(['00시', '05시', '06시', '07시', '08시', '09시', '10시', '11시', '12시',
       '13시', '14시', '15시', '16시', '17시', '18시', '19시', '20시', '21시',
       '22시', '23시', '04시', '01시', '02시', '03시'], dtype=object)

24시간 계속 시간 단위로 데이터가 수집되고 있는 것을 알 수 있다.

In [32]:
print("total 2019 : ", 24 * 151)
print("total 2020 : ", 24 * 152)

total 2019 :  3624
total 2020 :  3648


모든날, 모든 시간대에서 데이터가 수집되었다면  
2019년 기준 3624개가, 2020년 기준 3648개가 기록되어 있을 것이다.

In [33]:
stn_nm19.sort_values()

지평     1390
옥수     2401
달월     2677
백양리    2863
굴봉산    2874
       ... 
부천     3282
수원     3287
안산     3291
영등포    3291
부평     3322
Name: SBWY_STN_CD_NM, Length: 274, dtype: int64

In [34]:
stn_nm20.sort_values()

임진강      410
부산원동    1305
지평      1340
옥수      2407
달월      2603
        ... 
부천      3280
영등포     3294
정자      3297
안산      3303
부평      3340
Name: SBWY_STN_CD_NM, Length: 276, dtype: int64

수집 개수 순으로 정렬한 결과 두 데이터 모두 비슷하게 부천/수원/안산/부평/영등포가 상위에 있는 것을 알 수 있었다.   
여전히 의문인건, 그러면 수집이 되지 않은 날짜나 시간은 정녕 아무도 오가지 않은 것인가.

### 4-3. 시간대구분 (TMWD_DV)

어떤 시간대에서 가장 데이터가 많은지 봅시다.

In [35]:
sub19['TMWD_DV'].value_counts().sort_values()

03시        1
02시       29
01시     9660
04시    24356
00시    40121
       ...  
07시    41352
18시    41365
10시    41367
19시    41371
16시    41374
Name: TMWD_DV, Length: 24, dtype: int64

In [36]:
sub20['TMWD_DV'].value_counts().sort_values()

03시        3
02시       70
01시     5604
04시    24845
00시    40002
       ...  
07시    41682
18시    41727
19시    41755
16시    41755
10시    41759
Name: TMWD_DV, Length: 24, dtype: int64

두 데이터 다 새벽 이외에는 별 차이 없어보인다. 새벽 3시엔 1명 탄 경우도 있다. 와웅 ^^  
그런데 ... 새벽 3시에... 지하철이 다니던가...?

### [번외] 새벽 3시에 지하철을 운행하는 너.. 누구야!~

In [37]:
sub19[sub19['TMWD_DV'] == '03시']
#2019년 3월 26일.. 광운대역에서 ...하차 안했다. 둘다 결측값이다. 깜빡 속을뻔했다.

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV SBWY_ABRD_PRNB  \
476941  20190326              1019            광운대     03시             -    

       SBWY_GOFF_PRNB  
476941             -

In [38]:
sub20[sub20['TMWD_DV'] == '03시']
#2020년 데이터에서는 온양온천역/청량리역/성환역에서 승하차가 일어난 것을 볼 수 있다. 대체 뭐지?

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV SBWY_ABRD_PRNB  \
98477   20200118              1407           온양온천     03시             1    
119426  20200122              1014            청량리     03시             1    
513179  20200331              1725             성환     03시             -    

       SBWY_GOFF_PRNB  
98477              1   
119426             -   
513179             1

In [39]:
sub19[sub19['TMWD_DV'] == '02시']

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV SBWY_ABRD_PRNB  \
70279   20190113              1506            곤지암     02시             1    
162991  20190129              1866             상갈     02시             -    
198383  20190205              1001             서울     02시             -    
198427  20190205              1003             용산     02시             -    
198681  20190205              1015             회기     02시             -    
...          ...               ...            ...     ...            ...   
207675  20190206              1809             주안     02시             -    
208910  20190206              1907             가능     02시             -    
267121  20190217              1030             수서     02시             1    
457452  20190322              1808             동암     02시             -    
651482  20190425              1871           수원시청     02시             1    

       SBWY_GOFF_PRNB  
70279              1   
162991             -   
198383             -   
198427             -   
198681             -   
...               ...  
207675             -   
208910             -   
267121             -   
457452             -   
651482             -   

[29 rows x 6 columns]

In [40]:
sub20[sub20['TMWD_DV'] == '02시']

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV SBWY_ABRD_PRNB  \
12514   20200103              1269             강매     02시             -    
29438   20200106              1269             강매     02시             -    
52248   20200110              1269             강매     02시             -    
69188   20200113              1269             강매     02시             -    
74885   20200114              1269             강매     02시             -    
...          ...               ...            ...     ...            ...   
641367  20200423              1269             강매     02시             -    
663950  20200427              1269             강매     02시             -    
697697  20200503              1261          효창공원앞     02시             -    
782660  20200518              1269             강매     02시             -    
793990  20200520              1269             강매     02시             -    

       SBWY_GOFF_PRNB  
12514              -   
29438              -   
52248              -   
69188              -   
74885              -   
...               ...  
641367             -   
663950             -   
697697             -   
782660             -   
793990             -   

[70 rows x 6 columns]

강매역이 유난히 결측치로 많이 수집이 된 것 같다.  
속을 뻔 했다. 나중에 이런애들은 찬찬히 살펴볼 필요가 있는 것 같다.  
ex) 설연휴 연장운행

### 4-4. 승/하차 인원수 (ABRD_PRNB/GOFF_PRNB)

In [41]:
sub19

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV SBWY_ABRD_PRNB  \
0       20190101              1001             서울     00시             7    
1       20190101              1001             서울     05시            10    
2       20190101              1001             서울     06시            35    
3       20190101              1001             서울     07시            31    
4       20190101              1001             서울     08시            43    
...          ...               ...            ...     ...            ...   
857871  20190531              8014             일광     19시            95    
857872  20190531              8014             일광     20시            45    
857873  20190531              8014             일광     21시            70    
857874  20190531              8014             일광     22시            54    
857875  20190531              8014             일광     23시            50    

       SBWY_GOFF_PRNB  
0                 22   
1                  1   
2                  4   
3                 39   
4                101   
...               ...  
857871           109   
857872            73   
857873            57   
857874            29   
857875             6   

[857876 rows x 6 columns]

In [42]:
sub19.columns

Index(['RUN_DT', 'SBWY_STN_PRPR_NO', 'SBWY_STN_CD_NM', 'TMWD_DV',
       'SBWY_ABRD_PRNB', 'SBWY_GOFF_PRNB'],
      dtype='object')

In [46]:
sub19['SBWY_ABRD_PRNB']

0          7
1         10
2         35
3         31
4         43
          ..
857871    95
857872    45
857873    70
857874    54
857875    50
Name: SBWY_ABRD_PRNB, Length: 857876, dtype: Int64

In [47]:
sub20['SBWY_ABRD_PRNB']

0          4
1         13
2         59
3         32
4         61
          ..
860707    62
860708    41
860709    56
860710    13
860711    10
Name: SBWY_ABRD_PRNB, Length: 860712, dtype: Int64

In [44]:
print("\n**2019년도 광역(지하철) 승차 데이터**")
sub19['SBWY_ABRD_PRNB'] = prnb_to_int(sub19['SBWY_ABRD_PRNB'])
print("\n**2019년도 광역(지하철) 하차 데이터**")
sub19['SBWY_GOFF_PRNB'] = prnb_to_int(sub19['SBWY_GOFF_PRNB'])

print("")

print("\n**2020년도 광역(지하철) 승차 데이터**")
sub20['SBWY_ABRD_PRNB'] = prnb_to_int(sub20['SBWY_ABRD_PRNB'])
print("\n**2020년도 광역(지하철) 하차 데이터**")
sub20['SBWY_GOFF_PRNB'] = prnb_to_int(sub20['SBWY_GOFF_PRNB'])


**2019년도 광역(지하철) 승차 데이터**
' - ' count :  34319
'0' count :  0
after modification '0' count :  34319
after modification 0 count :  0
after modification 'nan' count :  34319
count    823557.000000
mean        393.157654
std         587.902559
min           1.000000
25%          71.000000
50%         192.000000
75%         470.000000
max        8662.000000
Name: SBWY_ABRD_PRNB, dtype: float64

**2019년도 광역(지하철) 하차 데이터**
' - ' count :  23400
'0' count :  0
after modification '0' count :  23400
after modification 0 count :  0
after modification 'nan' count :  23400
count    834476.000000
mean        392.113350
std         575.138181
min           1.000000
25%          65.000000
50%         194.000000
75%         481.000000
max        9796.000000
Name: SBWY_GOFF_PRNB, dtype: float64


**2020년도 광역(지하철) 승차 데이터**
' - ' count :  32038
'0' count :  0
after modification '0' count :  32038
after modification 0 count :  0
after modification 'nan' count :  32038
count    828674.000000
mean        285

비슷한 수치를 보인다. 이도 마찬가지로 각 데이터별 MAX값인 역을 찍어보았다.

In [48]:
#2019년 지하철 승차데이터 MAX ==> 아침 8시 "가산디지털단지"
sub19[sub19['SBWY_ABRD_PRNB'] == max(sub19[sub19['SBWY_ABRD_PRNB'] != np.nan]['SBWY_ABRD_PRNB'])]

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV  SBWY_ABRD_PRNB  \
791791  20190520              1702        가산디지털단지     08시            8662   

        SBWY_GOFF_PRNB  
791791             633

In [49]:
#2019년 지하철 하차데이터 MAX ==> 밤 10시, "구일역"
sub19[sub19['SBWY_GOFF_PRNB'] == max(sub19[sub19['SBWY_GOFF_PRNB'] != np.nan]['SBWY_GOFF_PRNB'])]

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV  SBWY_ABRD_PRNB  \
326930  20190227              1813             구일     22시             603   

        SBWY_GOFF_PRNB  
326930            9796

In [51]:
#2020년 지하철 승차데이터 MAX ==> 아침 8시, "가산디지털단지"
sub20[sub20['SBWY_ABRD_PRNB'] == max(sub20[sub20['SBWY_ABRD_PRNB'] != np.nan]['SBWY_ABRD_PRNB'])]

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV  SBWY_ABRD_PRNB  \
30675  20200106              1702        가산디지털단지     08시            8607   

       SBWY_GOFF_PRNB  
30675             592

In [53]:
#2020년 지하철 하차데이터 MAX ==> 저녁 6시, "경마공원"
sub20[sub20['SBWY_GOFF_PRNB'] == max(sub20[sub20['SBWY_GOFF_PRNB'] != np.nan]['SBWY_GOFF_PRNB'])]

RUN_DT  SBWY_STN_PRPR_NO SBWY_STN_CD_NM TMWD_DV  SBWY_ABRD_PRNB  \
104160  20200119              1451           경마공원     18시              58   

        SBWY_GOFF_PRNB  
104160            8660

---------

## DATA SET SUMMARY

1. 수집기간 : 2019년 1월 1일 ~ 2019년 5월 ~ 31일 / 2020년 1월 1일 ~ 2020년 5월 31일
2. 데이터 분류 : 여객(기차), 광역(지하철)
3. 데이터 변수 : 운행날짜, 역코드, 역이름, 승차인원, 하차인원 (지하철의 경우 시간대변수도 있음)
  
  
  
---

<h1> 5. 데이터셋 활용방안 </h1>

<h3> 1. 승하차 인원수로 히트맵 만들기! </h3>

<b> 지하철역 경위도 좌표를 수집하자! </b>
    
1-1. 여객(기차)
- 날짜별 승하차 인원수 기준 히트맵 생성.
        
1-2. 광역(지하철)
- 시간대별 승차/하차 인원수를 기준으로 히트맵 생성 (시간대 분리 예시 : 아침(출근길) / 점심 / 저녁(퇴근길) / 밤
     
    
 <h4><b>--> 기본적으로 사람이 많은, 특정 시간대에 몰리는 역의 경우 배차를 늘림으로써 사람을 분산, 더 안전하게 이동 가능</b></h4>
    
<br>
    

<h3>2. 나눈 시간대에 맞춰서 각 시간대별 승/하차 인원 TOP10 추출 </h3>

- 출근길/퇴근길/나른한오후/핫한새벽 별로 승하차인원이 많은 역이 모두 다 다를 것이다. 

- 그 역들의 특징을 알아보고, 시간대에 맞춰서 승하차인원을 조절해보자. (사실 위와 다를 건 크게 없다...ㅎ)
 
<br>

<h3>3. 워낙 쓸만한 변수 자체도 시간대/승하차인원 정도 밖에 없어서 할 수 있는건 이 정도가 아닐까!</h3>

<br><hr>

<h2> 6. 관련 외부 데이터 수집 </h2>

<h3> 1. [공공데이터포털] 코레일 지하철 주소 데이터 </h3>
    <ul>
        <li> 276개로, 코레일 소속 지하철 모든 지하철을 담고 있다고 볼 수 있다! </li>
        <li> 여기에 있는 도로명 주소를 이용해 x,y 좌표를 얻을 수 있다. </li>
    </ul> 

<h3> 2. [서울열린데이터광장] 서울시 지하철 호선별 역별 시간대별 승하차 인원 정보 </h3>
    <ul>
        <li> 호선별, 역별, 시간대별로 승하차인원 정보를 기록한 데이터로 추가자료로 이용할 수 있을 것 같다. </li>
    </ul>

### 6-1. 좌표 변환
[여기](http://blog.daum.net/geoscience/1436) 을 참고했다.

In [ ]:
import requests; from urllib.parse import urlparse
import pandas as pd
import geopandas

In [ ]:
# 카카오맵 주소 검색 API: https://developers.kakao.com/docs/restapi/local#주소-검색
address = "강원 원주시 단구로 171"
url = "https://dapi.kakao.com/v2/local/search/address.json?&query=" + address
result = requests.get(urlparse(url).geturl(),
                      headers={"Authorization":"KakaoAK 4d25f7ae6c4cab6bc96c746280e403af"})
json_obj = result.json()
print(json_obj)

In [ ]:
# 검색 결과 정렬
list = []
for document in json_obj['documents']:
    val = [document['road_address']['building_name'], document['address_name'], document['y'], document['x']]
    list.append(val)
    
df = pd.DataFrame(list, columns = ['building_name', 'address_name', 'lat', 'lon'])
df

In [61]:
ko_sub = pd.read_csv("추가데이터/공공데이터포털/한국철도시설공단_코레일_지하철_주소데이터_20191121..csv", encoding='cp949')

In [62]:
ko_sub

철도운영기관명   선명       역명                    지번주소                       도로명주소
0       코레일  1호선       가능      경기도 의정부시 가능동 197-1            경기도 의정부시 평화로 633
1       코레일  1호선  가산디지털단지     서울특별시 금천구 가산동 468-4           서울특별시 금천구 벚꽃로 309
2       코레일  1호선       간석       인천광역시 남동구 간석동 757        인천광역시 남동구 석정로 522-14
3       코레일  1호선       개봉       서울특별시 구로구 개보동 415         서울특별시 구로구 경인로40길 47
4       코레일  1호선       관악  경기도 안양시 만악구 석수동 101-16  경기도 안양시 만안구 경수대로 1273번길 46
..      ...  ...      ...                     ...                         ...
270     코레일   수인       월곶       경기도 시흥시 월곶동 992-8       경기도 시흥시 월곶중앙로 14번길 56
271     코레일   수인       인천                     NaN      인천 중구 차이나타운로26번길 12-17
272     코레일   수인     인천논현     인천광역시 남동구 논현동 731-4          인천광역시 남동구 청능대로 567
273     코레일   수인      인하대                     NaN               인천 남구 독배로 315
274     코레일   수인      호구포     인천광역시 남동구 논현동 731-1          인천광역시 남동구 호구포로 205

[275 rows x 5 columns]

In [63]:
# 주소 검색 함수 정의
def address_to_latlon(address):
    url = "https://dapi.kakao.com/v2/local/search/address.json?&query=" + address
    result = requests.get(urlparse(url).geturl(),
                          headers={"Authorization":"KakaoAK 4d25f7ae6c4cab6bc96c746280e403af"})
    json_obj = result.json()
    for document in json_obj['documents']:
        val = [document['road_address']['building_name'], document['address_name'], document['y'], document['x']]
    return val

In [ ]:
list = []
for address in ko_sub['도로명주소']:
    list.append(address_to_latlon(address))
    
df1 = pd.DataFrame(list, columns = ['building_name', 'address_name', 'lat', 'lon'])
df1.head()

## 6-2. Folium 시각화

## 7. 시간대별 나눠서 승/하차 TOP10 뽑아보기